[http1] streaming request bodies

deweerdt opened the pull request:
This commit introduces the infrastructure needed to stream request bodies
for http1, we have the two readers, chunked and content-length use the
h2o_req_t::write_req.cb to emit body chunks.

deweerdt commented:
Thank you @i110  and @kazuho I didn't realize the error failure was coming from the non-streaming case. Addressed. I've also addressed the issue pointed out by @kazuho above. I believe this can be cleaned up by merging some of the streaming infrastructure used by http1 and http2. I'll take a stab at that.

kazuho commented on lib/http1.c:
```diff
@@ -224,22 +233,21 @@ static int create_chunked_entity_reader(struct st_h2o_http1_conn_t *conn)
-    /* all input has arrived */
-    conn->req.entity = h2o_iovec_init(conn->sock->input->bytes + conn->_reqsize - reader->content_length, reader->content_length);
-    on_entity_read_complete(conn);
+    handle_one_body_fragment(conn, conn->sock->input->size, complete);
```
Doesn't this process excessive amount of input as request body if the last input consists of the end of the request body _and the first few bytes of the next request_ (i.e. pipelining)?

Please correct me if I'm incorrect, and if we have a test covering such case.

deweerdt replied:
That's right, i've added a test in fb8e95b that shows the previous code was broken. This is fixed now.

kazuho commented on lib/http1.c:
```diff
@@ -586,9 +648,18 @@ static void cleanup_connection(struct st_h2o_http1_conn_t *conn)
         return;
     }
 
+    if (conn->req.proceed_req) {
```
I'd appreciate it if you could add `!= NULL`. We apply operators so that the result would be a boolean, unless the input (being an integral value) is considered a boolean (due to lack of a built-in boolean type in C).

deweerdt replied:
Adressed in c3bcd0d thank you.

kazuho commented on lib/http1.c:
```diff
@@ -586,9 +648,18 @@ static void cleanup_connection(struct st_h2o_http1_conn_t *conn)
+    if (conn->req.proceed_req) {
+        conn->_req_entity_reader = NULL;
+        set_timeout(conn, 0, NULL);
+        h2o_socket_read_stop(conn->sock);
```
I'm not sure if this is the right moment. Don't we need to disarm the timeout and stop reading additional bytes every time when we pass something to the handler? Otherwise, the amount of data we buffer cannot be capped.

deweerdt replied:
Addressed in e4d894b

kazuho replied:
e4d894b looks good to me, but would you mind elaborating why we need to stop reading  (or consider the case of `conn->req.proceed_req` being non-NULL) in this function, after consulting the value of `conn->req.http1_is_persistent`?

It is my understanding that `cleanup_connection` is called only when the HTTP/1 stack successfully processes a request and sending a response. Assuming that is still the case, I _think_ `proceed_req` and `_req_entity_reader` must have been set to NULL, and reading from the socket should have been stopped. Maybe what we need are assertions.

deweerdt replied:
I've added assertions, thanks for the suggestion.

deweerdt replied:
> I've added assertions, thanks for the suggestion.

And the assertions found that 603c731 is necessary because `h2o_send_error` can complete the request without closing the connection.

deweerdt commented:
@kazuho i believe fa4fa74 addresses all your comments so far. I also have a test branch that integrates with @i110 's https://github.com/h2o/h2o/pull/2010 which looks good.

kazuho commented on lib/http1.c:
```diff
@@ -58,7 +58,7 @@ struct st_h2o_http1_conn_t {
     uint64_t _req_index;
     size_t _prevreqlen;
-    size_t _reqsize;
+    size_t _headers_size;
```
I would appreciate it if you could either change the name of the variable to something more appropriate, or add a comment explaining how the value is used.

This is because the value does not always represent the size of the header fields with the proposed change (when the request has a body, it becomes zero). It is my understanding that the amount of data left un-consumed in the socket read buffer.

deweerdt commented:
@kazuho i believe that 765027f addresses all you comments so far. A change that wasn't covered in the previous exchange is the connection close on the proxy 502s. Any thoughts on that?

kazuho commented:
> A change that wasn't covered in the previous exchange is the connection close on the proxy 502s. Any thoughts on that?

IIUC, the proposed change is to suggest closing the connection from proxy.c whenever it is impossible to connect to the origin. I am not sure if that is the correct thing to do.

At the moment, we close a H1 connection only when the framing becomes corrupt (i.e. when there would be a risk of splitting attack unless we close the connection). In case of an origin sending 502, the framing is not necessarily corrupt. It is beneficial to keep the connection open especially when only some of the requests are routed to an origin, while other requests being served by H2O itself.

I think we might want to consider the case you are trying to fix as part of #2010. The proxy handler returning 502 due to not being able to connect to an origin while the request from the client being inflight is a particular form of returning an early response. I am not sure of how we should handle the error, but my instinct is that it's something not unique to the proxy handler, but rather something general to all the handlers.

kazuho commented:
@deweerdt Regarding my previous comment, I asked @i110 on how he deals with the issue in H2, and his answer was than in #2010 the stream is closed by the `proceed_req` callback: https://github.com/h2o/h2o/blob/24d2036a5a9a3b7e218f8c82d3e92b5639760e65/lib/http2/connection.c#L657

I think this approach might be something we should adopt in H1 as well. WDYT?

deweerdt commented:
> I think this approach might be something we should adopt in H1 as well (or change to a different approach in both H1 and H2 code). WDYT?

That would work for me. Would you prefer a combined PR with #2007 and #2010 ?

kazuho commented:
> That would work for me. Would you prefer a combined PR with #2007 and #2010 ?

That's a good question. Yeah, let's tackle the issues independently. I think we can land the code here (that uses H2O_SEND_ERROR_HTTP1_CLOSE_CONNECTION) as-is, and fix it later in #2010 or in a follow-up PR of #2010.

Using H2O_SEND_ERROR_HTTP1_CLOSE_CONNECTION has efficiency issues, but works perfect as a short-term solution.

deweerdt commented:
> That's a good question. Yeah, let's tackle the issues independently. I think we can land the code here (that uses H2O_SEND_ERROR_HTTP1_CLOSE_CONNECTION) as-is, and fix it later in #2010 or in a follow-up PR of #2010.

That works. There's code more or less ready anyway, since i've tested an earlier version of this PR+2010 to help review #2010.

kazuho commented:
Thank you for working on this complex PR. Merged to master.
